### ISSUE TITLE
CCKD64 support needed for cckddump/cckdload utils

### ISSUE THREAD

**Issue by Fish-Git** [SDL-Hercules-390/hyperion#167]
Hercules ships with two mainframe assembler utilities designed to allow users to dump an entire mainframe dasd volume to CCKD format for transfer to Hercules on their personal computer ([`cckddump.hla`](https://github.com/SDL-Hercules-390/hyperion/blob/master/util/cckddump.hla)) and vice versa ([`cckdload.hla`](https://github.com/SDL-Hercules-390/hyperion/blob/master/util/cckdload.hla)). Both are HLASM assembler source files living in the repository's `'util'` subdirectory.

They were written by Greg Smith (the original author/designer of our CKD/CCKD and FBA/CFBA emulated dasd support) and were later updated by Chistopher Varlet in 2014 to support "very large dasds" (i.e. those with more than 64K tracks).

_**HOWEVER...**_ Even though the cckddump/cckdload programs both now support volumes with more than 64K tracks thanks to Chistopher Varlet, they both currently only support the original 32-bit CCKD (Compressed CKD) file format.

This means if a "large" dasd volume containing more than approximately 75,000 tracks (i.e. more than 4GB of data when compressed, such as a "3390-9" or larger) cannot be _compressed_ into a Hercules CCKD dasd image file of _less_ than 4GB in size, then you're basically screwed. You can't transfer that volume from your mainframe over to Hercules on your personal computer because the size of the resulting CCKD file would be larger than the current 4GB CCKD limit. (The reverse is also true too: existing CCKD64 format dasd volumes on Hercules cannot currently be transferred to the mainframe either since the cckdload utility is currently designed to only supports the original 32-bit CCKD file format, not the new CCKD64 file format.)

Both of these utilities need to either be updated to support the new CCKD64 file format or else new CCKD64 versions of the utilities need to be created.
____
**Note:** This issue is a sub-issue (sub-project) of Issue #20: ["We need 64-bit CCKD support"](https://github.com/SDL-Hercules-390/hyperion/issues/20)

**Comment by Fish-Git** [comment]
_**HELP!**_

I'm stuck!&nbsp;`:(`

I've encountered an addressability error and I don't know how best to fix it.

In `cckdlo64.hla` (the CCKD64 version of [`cckdload.hla`](https://github.com/SDL-Hercules-390/hyperion/blob/master/util/cckdload.hla) that I'm writing; see .zip file below for source and listing) there are two fields in a work area. One called `l2tab` and the other called `tempbuf`.

The l2tab field hold the CCKD Level 2 lookup table: a table of 256 Level 2 entries where each entry contains the dasd image file offset of the track data and the size and length of the track data.

The tempbuf field appears to be an I/O buffer or work buffer of some sort where CCKD data is manipulated and/or accessed (read into and written from??).

The problem is, in the old non-CCKD64 version of the program (`cckdload.hla`), the start of the `tempbuf` was very close to the very end of available addressability to that work area, so even though the tempbuf field itself was 16384 bytes long, it was still addressable since it was the very last field in the work area.

The `l2tab` field, which is defined _before_ the `tempbuf` field, under the old non-CCKD64 version of the program, was only 2048 bytes in size (256 entries of 8 bytes each: 4 for the dasd image file offset of the track data and 2 + 2 bytes for the size and length of that track data).

But with the new CCKD64 file format, each l2tab entry is now _16 bytes_ in size (i.e. _twice_ as big): 8 bytes for the dasd image file offset of the track data, 2 + 2 bytes for the size and length of that track data, and 4 bytes of padding for alignment). That makes the size of the `l2tab` field 4096 bytes (256 * 16), thereby blowing addressability to the `tempbuf` field! (The tempbuf field is now unaddressable!)

No matter which field I place before the other, it makes the other one unaddressable!&nbsp;`:(`

I would _like_ to move the 16K `tempbuf` field completely out of the existing work area dsect, and into its own `STORAGE OBTAIN` area, but I'm not experienced enough with z/OS programming to know how to do that, and I suspect doing so would only end up requiring another register to address that working storage anyway, thereby increasing the risk of a bug as a result. (I'm not familiar enough with the program's register usage to determine how much trouble (how risky) doing so would be.)

So I need some help.&nbsp;`:(`

Can any of you that are more experienced at z/OS programming help me out please? Can any of you figure out an elegant, low risk way of solving this problem?

Here's a zip file containing the new source code and the PDF listing that shows the problem. Page 64 of the listing is where the problem is.

* [cckdlo64.zip](https://github.com/SDL-Hercules-390/hyperion/files/2711387/cckdlo64.zip)

I'll return to studying the code but I would _really appreciate_ any help any of you could provide!&nbsp;`:(`

Thanks!

&nbsp;
**p.s.** If you have a quality visual diff tool, you can compare my new `cckdlo64.hla` code against the existing `cckdload.hla` code and see that the changes that needed to be made were very minor. If you don't have a good visual diff tool, here's an ExamDiff Pro .htm report you can look at instead. 99% of the code is exactly identical.

* [`cckdlo64.hla` vs. `cckdload.hla`.zip](https://github.com/SDL-Hercules-390/hyperion/files/2711411/cckdlo64.hla.vs.cckdload.hla.zip)

Thanks.

**Comment by mcisho** [comment]
I'll have a look. It might be a couple of days before you hear anything from me, I have several trips to airports to deliver/collect people coming up. If anyone else can offer a quicker turn-around, please feel free to do so.

**Comment by wably** [comment]
@Fish-Git it looks like you basically have three choices.

1.  You could remove those large areas (L2TAB and TEMPBUF) out of the dsect completely; and obtain separate storage for those areas (via STORAGE OBTAIN), and then put word pointers in the dsect for those two areas.  You would have to load a register with their addresses when you need them, of course.  If you can't spare a register, then you could:

2. Use the long displacement instructions.  For example, using the code as currently assembled in the listing you provided in the .zip file above, you could address TEMPBUF this way:  

        LAY     Rx,TEMPBUF

even though it is beyond the normal base register range.  There are a number of long displacement instructions - I believe that they all end in -Y for the mnemonic - that you can use to manipulate the contents of TEMPBUF in this way.

3.  If you could somehow spare a register, I would simply use a 2nd base register on the USING for this dsect and be done with it.

If I was coding this, I'd probably choose number 1 above.  I can almost always find a register I can use temporarily without disrupting stuff, like R14 or R15, even if I have to load it a couple of times to restore its address after a macro invocation for example.

**Comment by Fish-Git** [comment]
_**(Doh!)**_ &nbsp;  `%-P`

I really should have taken the time to study the code a bit closer before panicking and crying for help. The solution was much easier than expected:

> 3. If you could somehow spare a register, I would simply use a 2nd base register on the USING for this dsect and be done with it.

_Yep!_

Reviewing the code where the problem was occurring, I discovered there was indeed a register that wasn't being used! So I simply added a second register to the USING statement and inserted a pair of instructions to set it up. Problem solved!

Sorry to waste your time folks! I'll go crawl back under my rock now and leave you guys alone.&nbsp;`:-/`

_(Sheesh, Fish! Can you possibly be more lame/incompetent?! Sigh)_

**Comment by Fish-Git** [comment]
I'm still working on this but it's taking me much longer than expected as the needed changes are more involved than I originally anticipated. (I'm having to change a lot of code to use 64-bit "grande" instructions in many places so it's going slow.)

I've temporarily removed "IN PROGRESS" status as I need to temporarily stop working on it to fix a Hercules QETH/OSA issue with my CTCI-WIN code that has cropped up. When I finish I'll return to plugging away on this issue again.

**Comment by Fish-Git** [comment]
I'm having real trouble finding the time to work on this issue, and would like to request help from somebody.

_PLEASE!_ &nbsp; &nbsp; (pretty please?)

It's bad enough that I am totally inexperienced with using the the new z/Architecture 64-bit instructions but I am also totally inexperienced with assembling, linking and testing these utilities on z/OS or similar guests.

It would be really great if someone else could take over this issue from me and finish it. I can send you what I've already got so far (which may or may not be right!).

Thanks.

**Comment by pjfarleyiii** [comment]
Fish,

I think I am pretty competent at HLASM after 43+ years of writing and debugging in it.  I have DL'd the zip files in the above comments to take a look and see what I can do for you.  If you have a more recent version, please post a zip and I will pick it up.

Peter

**Comment by Fish-Git** [comment]
_**THANK YOU, Peter!**_

I had already begun making the changes that I believe needed to be made, but got yanked away right in the middle of my effort.

My more recent versions of the two utilities in question (`cckddu64.hla` and `cckdlo64.hla` -- the cckd64 equivalents of the two existing 32-bit cckd utilities, `cckddump.hla` and `cckdload.hla`) are both in a half finished state, although I believe the "load" utility _might_ be mostly complete (I think). The "dump" utility however, I know for a fact is largely unfinished. I left off right in the middle of it, near line 778 ("get area for rewrites").

I've already updated the the DSECTS in both of them to the new larger CCKD64 variety with the new larger 64-bit fields (refer to both the 'C' language struct definitions in source file [`cckd64.h`](https://github.com/SDL-Hercules-390/hyperion/blob/master/cckd64.h) as well as our [`cckddasd.html`](https://sdl-hercules-390.github.io/html/cckddasd.html#filestructure) web page for what the new layouts look like), and I tried to bookmark all of the places where I needed to make my changes, but I'm not 100% sure I've identified them all. I _believe_ I have. But I'm not 100% sure.

I've attached my unfinished effort (along with a few other files) to this comment:

* [cckd64.hla.zip](https://github.com/SDL-Hercules-390/hyperion/files/2930058/cckd64.hla.zip)

You will see two different versions of the two source files in question: one starting with an underscore and one without. The two files that start with underscores are ones where I've tried to textually identify which of the source lines I had bookmarked. (I know of no portable way to send someone a source file with certain lines "bookmarked". What you see is my attempt at doing so.) The other files are of course the same thing but without the screwy boomark identifiers.

The two .html web pages are [ExamDiff Pro](https://www.prestosoft.com/edp_examdiffpro.asp) reports so you can more easily see visually what changes I've already made. I have no idea whether they're valid or not.

The best way to approach this is to first familiarize yourself with the new CCKD64 control block layouts as compared to the old 32-bit CCKD layouts, so you can have an idea of what code will need to be changed, and then go from there.

That's it I guess.

I want you to know, Peter, _I really, REALLY, **REALLY**_ appreciate your offer to help out!

_**Truly!**_

Contact me off list if you have any questions or concerns about anything.

_**Thanks!!**_&nbsp;`:)))`

**Comment by pjfarleyiii** [comment]
I picked up your more current zip file.  I'll get back to you after some review and study.

Peter

**Comment by Fish-Git** [comment]
> How do I contact you off list?

By email: &nbsp; http://www.softdevlabs.com/about.html

**Comment by Fish-Git** [comment]
@pjfarleyiii 

How's it going, Peter?

**Comment by nipesi** [comment]
Hi Fish, was there ever a conclusion on this issue?

/Niels

**Comment by Fish-Git** [comment]
(cc: @pjfarleyiii )

> Hi Fish, was there ever a conclusion on this issue?

As you can see, I asked Peter for a status update bac in April but did not receive a reply, so no, this issue has still not been resolved yet. &nbsp; `:(`

**Comment by pjfarleyiii** [comment]
Apologies for not replying to your request for status sooner.  Unfortunately some serious RL health issues in my family prevented me from continuing as far as I wanted to go on this project for some time now.  The RL issues have improved recently so I am hoping to get back to working on this again very soon.

I mostly finished a first pass at the load utility (without any actual testing yet, just assembly) and was working on the dump utility when I ran out of "spare" time.  I will continue working on the dump utility and let you know when I have at least a first pass of that code done.

Peter

**Comment by Fish-Git** [comment]
> ... serious RL health issues in my family ...

_(ouch!)_ &nbsp;Sorry to hear that, Peter.&nbsp;`:(`

I hope things get better soon.

> The RL issues have improved recently ...

Great! Let's hope they continue to improve!

> I mostly finished a first pass at the load utility (without any actual testing yet, just assembly) and was working on the dump utility when I ran out of "spare" time. I will continue working on the dump utility and let you know when I have at least a first pass of that code done.

Please, take your time! Real life issues _always_ come first!

All the best to you and yours, Peter!

**Comment by gcb4** [comment]
Had a go at cckddu64 and have dumped a model27 with apparent success....
Ipl'd ZZSA and browsed a dataset located at around cyl 32000. 
Updated source attached.
[cckddu64.zip](https://github.com/SDL-Hercules-390/hyperion/files/3591558/cckddu64.zip)



cckddump 0.3.1 starting                                                     
main workarea is at address 0x1ea17000, 24-bit workarea is at address 0x7000
VOLXXX vtoc opened                                                          
VOLXXX has 32760 cyls, 15 trks/cyl and 491400 total trks                    
VOLXXX vtoc has 18000 total dscbs                                           
storage obtained for vtoc area, addr 0x1ea1a000 size 2520000                
reading VOLXXX vtoc                                                         
VOLXXX vtoc closed                                                          
1642 datasets are on VOLXXX                                                 
storage obtained for dsn area, addr 0x1ec82000 size 125020                  
storage obtained for trk vector, addr 0x1eca1000 size 1965600               
storage released for vtoc area, addr 0x1ea1a000 size 2520000                
369113 tracks out of 491400 will be dumped                                  
storage obtained for VOLXXX i/o area, addr 0x9000 size 852480               
storage obtained for compression, addr 0x1ea1a000 size 57344                
persistent c environment created, handle=0x1ea18ee8                         
file SYSUT2 opened for output                                               
file SYSUT2 closed for output                                               
file SYSUT2 opened for update                                               
file SYSUT2 closed for update                                               
file SYSUT1 closed

**Comment by Fish-Git** [comment]
> Had a go at `cckddu64` and have dumped a model-27 with apparent success...

> Updated source attached.

THANK YOU, Peter!&nbsp;`:)`

Can you do me a small(?) personal favor please? _(or anyone else that knows)_:

Can you send to me _(**OFF LIST!** not here!)_ instructions on how to build and run `cckddu64.hla` under z/OS? I'd like to try your new `cckddu64.hla` for myself if possible, and I'm an _extreme_ z/OS newbie!

I'd really appreciate it. Thanks.

(and again, if anyone _else_ can explain it to me instead (rather than bother Peter with it), that would be nice too! Thanks!)

**Comment by pjfarleyiii** [comment]
> > Had a go at `cckddu64` and have dumped a model-27 with apparent success...
> 
> > Updated source attached.
> 
> THANK YOU, Peter! `:)`

Hi Fish,

I am not gcb4, so it wasn't me who wrote that comment,

gcb4, whoever you are, thank you for getting this done!

> Can you do me a small(?) personal favor please? _(or anyone else that knows)_:
> 
> Can you send to me _(**OFF LIST!** not here!)_ instructions on how to build and run `cckddu64.hla` under z/OS? I'd like to try your new `cckddu64.hla` for myself if possible, and I'm an _extreme_ z/OS newbie!
 
I sent Fish some JCL for ASMACLG offline to assemble, link and execute cckddu64, but I do not have any library containing the COMPRES2 executable needed at link time to combine with CCKDDU64 so I cannot try it on my system, and I am not sure how we need to code the SYSUT1 and SYSUT2 DD statements for the actual dump execution.

gcb4, can you further enlighten us about where we can find COMPRES2 and how to code SYSUT1 / 2 for dumping?

Regards,

Peter

**Comment by gcb4** [comment]
Hi Peter,
Greg Smith uploaded a file to the Hercules-390 Group, cckd.jcl.xmi-aug2007.gz, which provides all of the zlib and bzip2 routines as well as instructions on how to assemble, link and execute the Dump and Load utilities. 
Please make contact if you run into any problems getting set up.

Regards..Gordon

**Comment by pjfarleyiii** [comment]
Thanks for the file link Gordon.  I DL'd it and will get it installed on my system later this week.

Peter

**Comment by gcb4** [comment]
Think I've resolved the outstanding issues with cckddu64. I had commented out an instruction while debugging cckddu64 and overlooked uncommenting it once done! Also sorted out an issue with residual data in the freespace calculation. Have now dumped 4 different model-27's, all of which exceed 4GB and they come up clean when checked with cckdcdsk64.

Here's the updated source
[cckddu64.hla.zip](https://github.com/SDL-Hercules-390/hyperion/files/3611611/cckddu64.hla.zip)

Gordon

**Comment by Fish-Git** [comment]
> Think I've resolved the outstanding issues with cckddu64.

> Here's the updated source:
> 
> * [cckddu64.hla.zip](https://github.com/SDL-Hercules-390/hyperion/files/3611611/cckddu64.hla.zip)

_THANK YOU_, Gordon!&nbsp;`:)`

(and sorry about calling you Peter earlier)

I will give it a try as soon as I find time and get back to you. &nbsp;(currently working on something else right now.)

**Comment by MaRvIn054** [comment]
Hello, 
I'm interessted as well with this support. I can't see any commit on the cckddump/cckdload utility, even in the SDL fork. what is the last status about it? 
I had a look on the last source attached to this topic, and gave it a try on a few M27 but I've been unsuccesful so far. At a first look on the dump, most of the datasets on volume looks ok, but I'm getting some errors reading some datasets at some point of the dump. The cckdcdsk64 utility reports a lot of error and gets stuck in cpu loop before reaching the end. First errors are related to length out of bounds at tracks over 76k.
Best Regards,
MaRvIn054.

**Comment by Fish-Git** [comment]
@MaRvIn054 
@gcb4 
@pjfarleyiii 

> I'm interested in this support as well. I can't see any commit on the `cckddump`/`cckdload` utility, even in the SDL fork. What is the last status about it?

My apologies to everyone. &nbsp; Due to workload I had completely forgotten about this issue.

While I still have not had time to try/verify Gordon's new `cckddu64` utility yet, I will commit it anyway as I trust him that it's working okay. &nbsp; _(Thank you, Gordon! and my apologies for the delay!)_

I have no idea what the status of `cckdlo64` is however, but I'm guessing it's not used very much anyway so I'm not as concerned about it.

I will post again once Gordon's  new 64-bit CCKD64 `cckddu64` utility has been committed.

**Comment by Fish-Git** [comment]
Gordon's new 64-bit CCKD64 version of cckddump called `cckddu64.hla` has been committed to SDL Hyperion in commits 3110b28280951b43fb06461f47e6ad874432f0b6 and ccf3cb93c6982698e485d2a77f2eaaa64d6aa4d0.

Please give it a try and let us know whether it works okay for you.

Thanks!

**Comment by gcb4** [comment]
Thank you Fish.
@MaRvIn054 - can you confirm if cckdcdsk64 shows errors in your newly dumped model-27's immediately after creating them, or only after using them in Hercules?

I've tested a few more of my own dumps and they are all 'clean' initially, but after using them in Hercules, they also start reporting errors, specifically "cdevhdr inconsistencies found, code 0001" when checking with cckdcdsk64. Removing the -ro flag results in the free space being rebuilt and the disk is 'clean' again.
What's also noticeable, is that after using these model-27's in Hercules their size has grown significantly. E.g. a 5.4 GB disk has increased to 7.5 GB without actually creating or updating any of the datasets on the volume. 

Regards..Gordon

**Comment by Fish-Git** [comment]
@gcb4 - do me a favor and try again, but this time please run `cckdmap` both immediately after _creating_ your new dasds (i.e. _before_ running `cckdcdsk64` to verify they're starting out clean) and then again afterwards (i.e. _after_ running `cckdcdsk64`).

If `cckdmap` reports there exists some free space (newly created CCKD/CCKD64 dasds should _not_ have any free space!), then please run `cckdcomp64` to remove the free space, and then run `cckdmap`, `cckdcdsk64`, `cckdmap` again afterwards.

THEN try using the dasd with Hercules.

What you are reporting  (them growing by many GB when used but not updated) doesn't make any sense!

**Comment by gcb4** [comment]
Hi Fish,
There will normally be free space on a cckddump created volume - the unused balance of the final 16384 block written to the output dataset if I remember correctly. 
  
Which options do you suggest we use with cckdcdsk64? I normally include the -ro switch when verifying the disk. 

Regards..Gordon

**Comment by MaRvIn054** [comment]
Hello, 
@gcb4 - Yes, I confirm I get errors before trying to use them. Anyway, I always set initial dump in ReadOnly and set up shadow files rather than updating the dump itself. Following a sample of errors reported just after download : 
cckdcdsk64.exe -ro -3 VOL001.CCKD64
HHC02499I Hercules utility cckdcdsk64 - CCKD64 dasd image verification - version 4.2.1.9826-SDL-g9b5cbe3d
HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: Sep 10 2019 at 11:14:57
HHC00379I 0:0000 CCKD file VOL001.CCKD64: starting cckdcdsk64 level 3 (read-only)...
**HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[76207] offset 0x013D6720E8298991 len 64690 is out of bounds**
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[96475] offset 0x0202CDAE0E412325 len 21545 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[76199] offset 0x029B18A171167365 len 12243 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[96507] offset 0x02A63D2810A78C47 len 28147 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[96416] offset 0x03B3321EC0817A4D len 40083 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[76220] offset 0x03D1F171B8DFE072 len 38358 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[96317] offset 0x0431E829AB830D87 len 57507 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[76248] offset 0x04D8C23CECDD9823 len 47320 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[96275] offset 0x0562CA3FB94FCD37 len 47867 is out of bounds
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1517)
....
HHC00368W 0:0000 CCKD file VOL001.CCKD64: free space errors detected
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1697)
HHC00371W 0:0000 CCKD file VOL001.CCKD64: trk[831] offset 0x0000000000986349 len 25039 validation error
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1782)
HHC00369W 0:0000 CCKD file VOL001.CCKD64: trk[832] hdr error offset 0x000000000098C518: AFF65D0211
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1737)
HHC00371W 0:0000 CCKD file VOL001.CCKD64: trk[912] offset 0x0000000000B3757E len 33250 validation error
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1782)
HHC00369W 0:0000 CCKD file VOL001.CCKD64: trk[913] hdr error offset 0x0000000000B3F760: 0000004955
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1737)
HHC00371W 0:0000 CCKD file VOL001.CCKD64: trk[1120] offset 0x0000000000FB4CC4 len 30124 validation error
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1782)
HHC00369W 0:0000 CCKD file VOL001.CCKD64: trk[1121] hdr error offset 0x0000000000FBC270: 558ED4BA81
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1737)
HHC00371W 0:0000 CCKD file VOL001.CCKD64: trk[2404] offset 0x0000000001907ACF len 23906 validation error
HHC00007I Previous message from function 'cckd64_chkdsk' at cckdutil64.c(1782)
....

I've run a cckdmap but I not sure to understand why the offset reported by cckdmap is different from the offset shown on error, and also why a huge free comes in the tracks flow : 
HHC03042I l2       0x0032098D94  0x0000001000        4096  75264 - 75519
HHC03042I l2       0x00320A7A92  0x0000001000        4096  75520 - 75775
HHC03042I l2       0x00323478D5  0x0000001000        4096  75776 - 76031
**HHC03042I l2       0x003261764B  0x0000001000        4096  76032 - 76287**
HHC03042I l2       0x00326D05D9  0x0000001000        4096  76288 - 76543
HHC03042I l2       0x0032A20B5A  0x0000001000        4096  76544 - 76799
HHC03042I l2       0x0032D84A95  0x0000001000        4096  76800 - 77055
....
HHC03042I l2       0x0038335318  0x0000001000        4096  96000 - 96255
HHC03042I l2       0x00383B4379  0x0000001000        4096  96256 - 96511
**HHC03042I free     0x00383B5379  0x000004AB8B      306059**
HHC03042I l2       0x00383FFF04  0x0000001000        4096  96512 - 96767
HHC03042I l2       0x003844DECD  0x0000001000        4096  96768 - 97023
HHC03042I l2       0x00384BCFAE  0x0000001000        4096  97024 - 97279
...

Best Regards,
MaRvIn054.

**Comment by gcb4** [comment]
@MaRvIn054 There were issues with the first version I released in this thread. To confirm that you are running the latest, line 1195 in the source must NOT be commented out. 

CCKDMAP should show free_num =1 and it should be the last entry before eof, and should be less than 16384.

Could you share your assemble and link-edit JCL and the output from the CCKDDU64 job? 

Regards..Gordon

**Comment by MaRvIn054** [comment]
@gcb4 : I can share the assembly output, will have to find how to attach a file, but I confirm I used the last version from this topic which matches the last comit in this GitHub (checked again). Execute on line 1195 (copy the rest) is not commented out.
Following the output of the dump utility :
cckddump 0.3.1 starting
main workarea is at address 0x15530000, 24-bit workarea is at address 0x7000
VOL001 vtoc opened
VOL001 has 32760 cyls, 15 trks/cyl and 491400 total trks
VOL001 vtoc has 15000 total dscbs
storage obtained for vtoc area, addr 0x15533000 size 2100000
reading VOL001 vtoc
VOL001 vtoc closed
370 datasets are on VOL001
storage obtained for dsn area, addr 0x15734000 size 28348
storage obtained for trk vector, addr 0x1573b000 size 1965600
storage released for vtoc area, addr 0x15533000 size 2100000
367607 tracks out of 491400 will be dumped
storage obtained for VOL001 i/o area, addr 0x9000 size 852480
storage obtained for compression, addr 0x15533000 size 57344
persistent c environment created, handle=0x15531ee8
file SYSUT2 opened for output
file SYSUT2 closed for output
file SYSUT2 opened for update
file SYSUT2 closed for update
file SYSUT1 closed 

Regards,
MaRvIn054.

**Comment by gcb4** [comment]
Ok thanks for confirming the source MaRvIn054. Please try and send your assemble and link-edit when you get a chance.

There's nothing unusual in the job output. The offsets reported as 'out of bounds' by cckdcdsk64 are clearly corrupt as they are way too big. I'll go through the code again to see if I can spot anything that might corrupt the l2 offsets, however this could take some time.

Not sure if this could be relevant, but which os version and hardware are you running cckddu64 on?

Regards..Gordon

**Comment by Fish-Git** [comment]
> Which options do you suggest we use with cckdcdsk64?

`-3 -ro` if you're just looking, `-3` if you actually want to fix.

**Comment by Fish-Git** [comment]
@MaRvIn054 

> Yes, I can share the assembly output. I will have to find how to attach a file though

Attaching a file to your GitHub comment/reply is easy: just drag & drop! (or copy & paste)

Note however that only certain file types are supported:

* ["File attachments on issues and pull requests"](https://help.github.com/en/github/managing-your-work-on-github/file-attachments-on-issues-and-pull-requests)

So if it's a text file (.txt), just drag and drop it into your comment. Otherwise convert it to a .zip file first, and then drag and drop the .zip file into your comment.

**Comment by Fish-Git** [comment]
@MaRvIn054 

(cckdcdsk64):
> HHC00365W 0:0000 CCKD file VOL001.CCKD64: trk[76207] offset 0x013D6720E8298991 len 64690 is out of bounds

(cckdmap):
> HHC03042I l2       0x003261764B  0x0000001000        4096  76032 - 76287

Let's take a look at track 76207 with `cckddiag64`.

Please do a:

```
cckddiag64  -d -c -1 -2  -t  -r 76207  -g   VOL001.CCKD64
```

and post the output here, either inline or as an attachment.

Thanks.

**Comment by gcb4** [comment]
Fish, thanks for jumping in here. 

While diagnosing one of my volumes a while ago, I picked up a bug in cckddiag64. Not sure if I need to open a new issue for it or can deal with it here?

When using the -o switch, and the offset in question exceeds 4GB then cckddiag64 only returns the 32-bit value of the offset:-

D:\Download\Hercules\Herc421w64>cckddiag64 -o **0x014A8F8FDC** 128 D:\Download\Hercules\devices\volxxx.3390
HHC02499I Hercules utility cckddiag64 - CCKD64/CFBA64 diagnostic program - version 4.2.1.9826-SDL-g9b5cbe3d
HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: Sep 10 2019 at 11:14:57


HHC02613I IMAGE offset 1250922460 (**0x000000004A8F8FDC**); length 128 (0x00000080) bytes:
+0000 C919698C DCBF36C0 72A7F143 915EF7E9  I......{.x1.j;7Z
+0010 4661FF44 DAD828B9 BD9D5E74 006C5534  ./...Q..].;..%..
+0020 1F2A435F E7D8423F BC6B05CE A50EE76E  ...^XQ...,..v.X>
+0030 47D6A0FC EED9070E AE8067E9 6B78CBC8  .O...R.....Z,..H
+0040 139C79EC 154E2E2C 57EE6FE5 1246F933  ..`..+....?V..9.
+0050 CE7515B4 522F6376 229A970B BBDAF24F  ..........p...2.
+0060 9998FFB2 FA3B6E6C 414670EC 011FF1F8  rq....>%......18  

Regards..Gordon

**Comment by Fish-Git** [comment]
> ... I discovered a bug in `cckddiag64`. Not sure if I need to open a new issue for it or if it can be dealt with here?

Normally I would say to open a new issue, but after reading your description of it, I don't see the need to do that in this case. The bug is obvious and I should have it fixed within minutes. Thanks for reporting it!

**Comment by Fish-Git** [comment]
> The bug is obvious and I should have it fixed within minutes.

Fixed in commit 866c42a498fc656eae0ab30d65e757fa8f9f629c.

**Comment by gcb4** [comment]
Nice quick job Fish!
> 
> 
> @gcb4 - do me a favor and try again, but this time please run `cckdmap` both immediately after _creating_ your new dasds (i.e. _before_ running `cckdcdsk64` to verify they're starting out clean) and then again afterwards (i.e. _after_ running `cckdcdsk64`).
> 
> If `cckdmap` reports there exists some free space (newly created CCKD/CCKD64 dasds should _not_ have any free space!), then please run `cckdcomp64` to remove the free space, and then run `cckdmap`, `cckdcdsk64`, `cckdmap` again afterwards.
> 

Here are the cckdmap reports of before and after cckdcomp64. There was no difference before and after cckdcdsk64.
[1. volxxx.cckdmap.initial.txt](https://github.com/SDL-Hercules-390/hyperion/files/3754462/1.volxxx.cckdmap.initial.txt)
[5. volxxx.cckdmap.after_comp64.txt](https://github.com/SDL-Hercules-390/hyperion/files/3754464/5.volxxx.cckdmap.after_comp64.txt)

> THEN try using the dasd with Hercules.
> 

Behaviour in Hercules is much better when using cckddu64/cckdcomp64 volumes than when using the cckddu64 volumes directly. No more "cckd64_chk_space" messages flooding the log.  

> What you are reporting (them growing by many GB when used but not updated) doesn't make any sense!

Here's a cckdmap of the same volume created with cckddu64 and used directly in Hercules. It looks like the L2 entries were being moved to the beginning of the file? Can't see the largest free block in the report though. Running this file through cckdcomp64 resulted in the 2GB being released.

[volxxx.old.txt](https://github.com/SDL-Hercules-390/hyperion/files/3754489/volxxx.old.txt)

Regards...Gordon

**Comment by MaRvIn054** [comment]
Hello, 
I've made a new test with another volume, and I'm still getting the same behaviour : 
- [chkdsk64_VOL000.txt](https://github.com/SDL-Hercules-390/hyperion/files/3755809/chkdsk64_VOL000.txt)
- [Map_VOL000_CCKD64.txt](https://github.com/SDL-Hercules-390/hyperion/files/3755810/Map_VOL000_CCKD64.txt)
- [ckddiag_VOL000_49418.txt](https://github.com/SDL-Hercules-390/hyperion/files/3755812/ckddiag_VOL000_49418.txt)
- [ckddiag_VOL000_17103.txt](https://github.com/SDL-Hercules-390/hyperion/files/3755814/ckddiag_VOL000_17103.txt)

The highly suspect area is the huge free space rooms coming randomly accros the dump (cf map) : 
...
HHC03042I l2       0x000BB5F90C  0x0000001000        4096  16896 - 17151
HHC03042I free     0x000BC58D37  0x000010ABEC     1092588
HHC03042I l2       0x000BD63923  0x0000001000        4096  17152 - 17407
...
HHC03042I l2       0x001BB73531  0x0000001000        4096  41728 - 41983
HHC03042I free     0x001BDF2871  0x00000D617E      876926
HHC03042I l2       0x001BEC89EF  0x0000001000        4096  41984 - 42239
...
HHC03042I l2       0x002D629B7D  0x0000001000        4096  58880 - 59135
HHC03042I free     0x002D996CF9  0x00000006DF        1759
HHC03042I l2       0x002DBC7367  0x0000001000        4096  59136 - 59391
...
HHC03042I l2       0x0043561D8C  0x0000001000        4096  93696 - 93951
HHC03042I free     0x0043562D8C  0x00000B22B2      729778
HHC03042I l2       0x004361503E  0x0000001000        4096  93952 - 94207
HHC03042I free     0x00438682F2  0x0000001255        4693
HHC03042I free     0x0043A782FD  0x000000281E       10270
HHC03042I l2       0x0043D71E3C  0x0000001000        4096  94208 - 94463
HHC03042I free     0x0043FF835D  0x0000001769        5993
HHC03042I l2       0x0044162A27  0x0000001000        4096  94464 - 94719
...

Best Regards,
MaRvIn054.

**Comment by Fish-Git** [comment]
It's been almost a year and a half since any comments were made to this issue so I'm willing to close it for now.

Hercules's existing `cckddu64.hla` _seems_ to work (mostly?) fine.

_PLUS,_ there _are_ other methods to achieve transferring "dasds" from real mainframes to Hercules, not the least of which is via FTPing the individual datasets and using Hercules's `dasdload` utility to add them to your Hercules constructed dasd (as mentioned in[ message <span>#</span>2257](https://hercules-390.groups.io/g/group/message/2257) of the following Hercules support group thread):

* [Hercules support group thread "**CCKDDUMP**"](https://hercules-390.groups.io/g/group/message/2216)

(inline copy of message [<span>#</span>2257](https://hercules-390.groups.io/g/group/message/2257)):
> After a private exchange here a small procedure a reasonable number of files:
> 
> Assuming enough space
> 
> **`XMIT N1.BENUTZER DA('whatever.pds') outdsname(whatever.pds.xmi)`**
> 
> or any sequential file.
> 
> transfer these things to the workstation using either FTP (binary)  in one command `GET *` or via `IND$file`
> 
> Then on the workstation create a disk with the hercules `dasdinit` and then use `dasdload` 
> 
>  add the disk to your hercules  config
> 
> fire up MVS
> 
> You may either want to catalog the datasets or copy them to another disk
> 
> Have fun
> 
> Peter Sylvester
